libnetwork/idm/idm.go

```diff
@@ -4,40 +4,36 @@ package idm
 import (
 	"errors"
 	"fmt"
+	"sync"
 
-	"github.com/docker/docker/libnetwork/bitseq"
-	"github.com/docker/docker/libnetwork/datastore"
+	"github.com/docker/docker/libnetwork/bitmap"
 )
 
-// Idm manages the reservation/release of numerical ids from a contiguous set
+// Idm manages the reservation/release of numerical ids from a contiguous set.
+// Idm is safe for concurrent use.
 type Idm struct {
 	start  uint64
 	end    uint64
-	handle *bitseq.Handle
+	handle *bitmap.Bitmap
+	mu     sync.Mutex
 }
```

Comment:
Convention is to put the mutex above the fields that it protects; in this case (from the code) I assume `handle` doesn't need protection, but `start` and `end` do, so perhaps we should add a comment about that, and move the order of fields;

```suggestion
type Idm struct {
	handle *bitmap.Bitmap

	// mu protects start and end.
	mu         sync.Mutex
	start, end uint64
}
```

Reply:
Oh, hm.. I'm slightly confused now, because I see locks also before `i.handle.Set()`. I guess in either case we would need some comment 🤔 (and possibly some of the lock/unlocks earlier, but not sure)

Reply:
👆 hm.. right, so `i.start` and `i.end` are never updated (only set on `New()`), so should the mutex actually not be on `Idm`, but inside `bitmap.Bitmap` ?

Reply:
> so should the mutex actually not be on `Idm`, but inside `bitmap.Bitmap` ?

No. Consumers shouldn't have to pay for synchronization they don't need. That also goes for consumers which do need to be safe for concurrent use: they may need to call multiple methods inside critical sections, in which case any synchronization within bitmap will also be unnecessary.
https://github.com/moby/moby/blob/4ee0cf6878a076bee11b4cc71751e29d01935b59/libnetwork/bitmap/sequence.go#L33